README.md
# FixFloat

This fixed float math library provides an efficient and deterministic solution for arithmetic and geometric operations in Unity projects. The library is designed to be compatible with Unity's Burst Compiler, ensuring high-performance and low-overhead when used in conjunction with Unity's job system.


## Features

- Fixed-point arithmetic using FixNumber for deterministic calculations

- 2D fixed-point vector operations with FixVec

- Trigonometric functions using fixed-point angles

- Rotation matrix support with RotationMatrix

- Angle conversion and manipulation utilities in FixAngle

- Unity and Burst Compiler compatible


## Installation

To use iShape.FixFloat in your Unity project, follow these steps:

- Open your Unity project.
- In the top menu, select "Window" > "Package Manager".
- Click on the "+" button in the top-left corner of the Package Manager window.
- Select "Add package from git URL...".
- Enter the following URL: https://github.com/iShapeUnity/FixFloat.git
- Click the "Add" button.
- Wait for the package to be imported.
- In your C# scripts, add the following using statement to access the library:

```csharp
using iShape.FixFloat;
```


## How It Works

The \`**FixNumber**\` class uses fixed-point arithmetic to perform calculations with a high degree of precision and determinism. It supports numbers in the range 2^21 - 1 to -2^21 + 1 with a precision of 1/1024, and is most suitable for the range 10,000,000 to -10,000,000 with a precision of 0.01.

Fixed-point numbers are represented using a fixed number of bits for the fractional part. In this implementation, the number of bits representing the fractional part of the fixed-point number is 10, which allows for a precision of 1/1024 or approximately 0.001.

Here are some examples of fixed-point number representation:

- 1 / 1024 ≈ 0.001 (represented as 1)
- 256 / 1024 = 0.25 (represented as 256)
- 1024 / 1024 = 1 (represented as 1024)
- (1024 + 512) / 1024 = 1.5 (represented as 1536)
- (2048 + 256) / 1024 = 2.25 (represented as 2304)

By using the \`**FixNumber**\` class, you can perform arithmetic operations using long values while maintaining the precision of floating-point numbers, ensuring deterministic behavior across different platforms and devices.


## Usage

### FixNumber

The \`**FixNumber**\` class represents a fixed-point number using a \`**long**\` as the underlying storage, allowing deterministic arithmetic operations. Use \`**FixNumber**\` for calculations instead of \`**float**\` or \`**double**\` when deterministic behavior is required. \`**FixNumber**\` provides a way to perform arithmetic operations with \`**long**\` values while maintaining the precision of floating-point numbers.

```csharp
long a = 3.14f.ToFix();
long b = 2.0f.ToFix();
long result = a * b;
float resultAsFloat = result.ToFloat();
```

### FixVec

The \`**FixVec**\` struct represents a 2D fixed-point vector, providing various utility methods and operators for vector operations. Use \`**FixVec**\` for 2D geometric calculations when deterministic behavior is required.

```csharp
FixVec vec1 = new FixVec(1.0f.ToFix(), 2.0f.ToFix());
FixVec vec2 = new FixVec(3.0f.ToFix(), 4.0f.ToFix());

FixVec sum = vec1 + vec2;
FixVec difference = vec1 - vec2;
```

### FixAngle
The \`**FixAngle**\` class provides various utility methods for working with fixed-point angles, including trigonometric functions and angle conversion.


```csharp
long angle = (Mathf.PI / 2).ToFix().RadToFixAngle();
long sin = angle.Sin();
long cos = angle.Cos();
```


## Compatibility

This library is designed to be compatible with Unity and the Burst Compiler, enabling high-performance arithmetic and geometric operations in Unity projects that utilize the job system and Burst.


## License

This Fixed Float Math Library is released under the [MIT license](https://en.wikipedia.org/wiki/MIT_License).




CHANGELOG.md
# Change Log

## [0.0.1] - 2023-04-28

### Added

- fixed float math
- fixed float vec math
- fixed float angle math

Tests/Runtime/iShape/FixFloat/FixVecTests.cs
using iShape.FixFloat;
using NUnit.Framework;
using Unity.Mathematics;

namespace Tests.Runtime.iShape.FixFloat {

    public class FixVecTests {

        [Test]
        public void TestMath() {
            var v0 = new FixVec(-1000, 1000) + new FixVec(1000, -1000);
            Assert.AreEqual(v0, new FixVec(0, 0));

            var v1 = new FixVec(-1000, 0) - new FixVec(-1000, -1000);
            Assert.AreEqual(v1, new FixVec(0, 1000));

            var a0 = new FixVec(-512, 0).SqrDistance(new FixVec(512, 0));
            Assert.AreEqual(a0, FixNumber.Unit);
        }
        
        [Test]
        public void TestSQRLengthDebug_0() {
            var v0 = new FixVec(70034108, 654178);
            var v1 = v0.ToFloat2();
            
            var s0 = v0.SqrLength.ToFloat();
            var s1 = v1.SqrLength();
            
            Assert.AreEqual(s0, s1, 0.001f * s1);
        }
        
        [Test]
        public void TestSQRLengthDebug_1() {
            var v0 = new FixVec(12, 1);
            var v1 = v0.ToFloat2();
            
            var s0 = v0.SqrLength.ToFloat();
            var s1 = v1.SqrLength();
            
            Assert.AreEqual(s0, s1, 0.01 * (s1 + s0 + 1));
        }
        
        [Test]
        public void TestLengthDebug() {
            var v0 = new FixVec(1024 * 1024, 0);
            var v1 = v0.ToFloat2();
            
            var s0 = v0.Length.ToFloat();
            var s1 = v1.Length();
            
            Assert.AreEqual(s0, s1, 0.01 * (s1 + s0 + 1));
        }
        
        [Test]
        public void TestNormalizeDebug() {
            var v0 = new FixVec(12, 1);
            var v1 = v0.ToFloat2();
            
            var s0 = v0.Normalize.ToFloat2();
            var s1 = v1.Normalize();
            
            Assert.AreEqual(s0.x, s1.x, 0.05);
            Assert.AreEqual(s0.y, s1.y, 0.05);
        }
        
        [Test]
        public void TestSQRLength() {
            long x = 12;
            while (x < 25_000) {
                long y = 1;
                while (y < 10_000) {
                    var v0 = new FixVec(x, y);
                    var v1 = v0.ToFloat2();
            
                    var s0 = v0.SqrLength.ToFloat();
                    var s1 = v1.SqrLength();
                
                    Assert.AreEqual(s0, s1, 0.05 * (s1 + s0 + 1));
					
                    y += 17;
                }
                x += 13;
            }
        }
        
        [Test]
        public void TestLength() {
            long x = 12;
            while (x < 10_000) {
                long y = 1;
                while (y < 100_000) {
                    var v0 = new FixVec(x, y);
                    var v1 = v0.ToFloat2();
            
                    var s0 = v0.Length.ToFloat();
                    var s1 = v1.Length();
                
                    Assert.AreEqual(s0, s1, 0.05 * (s1 + s0 + 1));
					
                    y += 131;
                }
                x += 109;
            }
        }
        
        [Test]
        public void TestNormalize() {
            long x = 12;
            while (x < 10_000) {
                long y = 1;
                while (y < 100_000) {
                    var v0 = new FixVec(x, y);
                    var v1 = v0.ToFloat2();
            
                    var s0 = v0.Normalize.ToFloat2();
                    var s1 = v1.Normalize();

                    var message = "(" + x + ", " + y + ")"; 
                    
                    Assert.AreEqual(s0.x, s1.x, 0.05, message);
                    Assert.AreEqual(s0.y, s1.y, 0.05, message);
					
                    y += 131;
                }
                x += 109;
            }
        }
        
        [Test]
        public void TestOrtho() {
            AssertOrtho(new float2(0, 1), new float2(1, 0));
            AssertOrtho(new float2(0, 10), new float2(1, 0));
            AssertOrtho(new float2(0, 100), new float2(1, 0));

            AssertOrtho(new float2(0.1f, 0.9f), new float2(0.9f, -0.1f).Normalize());
            AssertOrtho(new float2(1, 9), new float2(0.9f, -0.1f).Normalize());
            AssertOrtho(new float2(10, 90), new float2(0.9f, -0.1f).Normalize());

            AssertOrtho(new float2(1, 0), new float2(0, -1));
            AssertOrtho(new float2(10, 0), new float2(0, -1));
            AssertOrtho(new float2(100, 0), new float2(0, -1));

            AssertOrtho(new float2(0, -1), new float2(-1, 0));
            AssertOrtho(new float2(0, -10), new float2(-1, 0));
            AssertOrtho(new float2(0, -100), new float2(-1, 0));

            AssertOrtho(new float2(-1, 0), new float2(0, 1));
            AssertOrtho(new float2(-10, 0), new float2(0, 1));
            AssertOrtho(new float2(-100, 0), new float2(0, 1));

            var r = 0.01f;
            var hp = 0.5f * math.PI;
            while (r < 2000) {
                var a = -0.001f;
                while (a < 2 * math.PI) {
                    var v = r * new float2(math.cos(a), math.sin(a));
                    var n = new float2(math.cos(a - hp), math.sin(a - hp));

                    AssertOrtho(v, n);
                    a += 0.1f;
                }

                r *= 5;
            }
        }
        
        private void AssertOrtho(float2 v, float2 m) {
            var f = v.ToFixVec();
            var n = m.ToFixVec();

            AssertVec(f.Ortho(clockwise: true), n);
            AssertVec(f.Ortho(clockwise: false), n.Reverse);
        }

    
        private void AssertVec(FixVec a, FixVec b) {
            Assert.AreEqual(a.x, b.x, 130);
            Assert.AreEqual(a.y, b.y, 130);
        }
    }
}

Tests/Runtime/iShape/FixFloat/FixAngleTests.cs
using iShape.FixFloat;
using NUnit.Framework;
using Unity.Mathematics;

namespace Tests.Runtime.iShape.FixFloat {

    public class FixAngleTests {

        [Test]
        public void TestMul() {
            Assert.AreEqual(0d.ToFixAngle(), 0);
            Assert.AreEqual((0.125 * math.PI).ToFixAngle(), 64);
            Assert.AreEqual((0.25 * math.PI).ToFixAngle(), 128);
            Assert.AreEqual((0.375 * math.PI).ToFixAngle(), 192);
            Assert.AreEqual((0.5 * math.PI).ToFixAngle(), 256);
            Assert.AreEqual((0.625 * math.PI).ToFixAngle(), 320);
            Assert.AreEqual((0.75 * math.PI).ToFixAngle(), 384);
            Assert.AreEqual((0.875 * math.PI).ToFixAngle(), 448);
            Assert.AreEqual((1.0 * math.PI).ToFixAngle(), 512);
            Assert.AreEqual((1.125 * math.PI).ToFixAngle(), 576);
            Assert.AreEqual((1.25 * math.PI).ToFixAngle(), 640);
            Assert.AreEqual((1.375 * math.PI).ToFixAngle(), 704);
            Assert.AreEqual((1.5 * math.PI).ToFixAngle(), 768);
            Assert.AreEqual((1.625 * math.PI).ToFixAngle(), 832);
            Assert.AreEqual((1.75 * math.PI).ToFixAngle(), 896);
            Assert.AreEqual((1.875 * math.PI).ToFixAngle(), 960);
            Assert.AreEqual((2.0 * math.PI).ToFixAngle(), 1024);

            Assert.AreEqual(0.ToFix().AngleToFixAngle(), 0);
            Assert.AreEqual(45.ToFix().AngleToFixAngle(), 128);
            Assert.AreEqual(90.ToFix().AngleToFixAngle(), 256);
            Assert.AreEqual(135.ToFix().AngleToFixAngle(), 384);
            Assert.AreEqual(180.ToFix().AngleToFixAngle(), 512);
            Assert.AreEqual(225.ToFix().AngleToFixAngle(), 640);
            Assert.AreEqual(270.ToFix().AngleToFixAngle(), 768);
            Assert.AreEqual(315.ToFix().AngleToFixAngle(), 896);
            Assert.AreEqual(360.ToFix().AngleToFixAngle(), 1024);


            Assert.AreEqual(0d.ToFix().RadToFixAngle(), 0);
            Assert.AreEqual(0d.ToFix(), 0);

            Assert.AreEqual((0.5 * math.PI).ToFix().RadToFixAngle(), 256, 1);
            Assert.AreEqual((0.5 * math.PI).ToFix(), 1608);

            Assert.AreEqual((1.0 * math.PI).ToFix().RadToFixAngle(), 512, 1);
            Assert.AreEqual((1.0 * math.PI).ToFix(), 3216);

            Assert.AreEqual((1.5 * math.PI).ToFix().RadToFixAngle(), 768, 1);
            Assert.AreEqual((1.5 * math.PI).ToFix(), 4825);

            Assert.AreEqual((2.0 * math.PI).ToFix().RadToFixAngle(), 1024, 1);
            Assert.AreEqual((2.0 * math.PI).ToFix(), 6433);
        }
		
        [Test]
        public void TestCos() {
            var rad = -100d;

            while (rad < 100) {
                var angle = 180 * rad / math.PI;
                var fixAngle0 = rad.ToFix().RadToFixAngle();

                var fixFloat = angle.ToFix();
                var fixAngle1 = fixFloat.AngleToFixAngle();

                Assert.AreEqual(fixAngle0, fixAngle1, 1);

                var cs = math.cos(rad).ToFix();
                var cs0 = fixAngle0.Cos();
                var cs1 = fixAngle1.Cos();

                Assert.AreEqual(cs, cs0, 8);
                Assert.AreEqual(cs, cs1, 8);

                rad += 0.001;
            }
        }
        
        [Test]
        public void TestSin() {
            var rad = -100d;

            while (rad < 100) {
                var angle = 180 * rad / math.PI;
                var fixAngle0 = rad.ToFix().RadToFixAngle();

                var fixFloat = angle.ToFix();
                var fixAngle1 = fixFloat.AngleToFixAngle();

                Assert.AreEqual(fixAngle0, fixAngle1, 1);

                var sn = math.sin(rad).ToFix();
                var sn0 = fixAngle0.Sin();
                var sn1 = fixAngle1.Sin();

                Assert.AreEqual(sn, sn0, 8);
                Assert.AreEqual(sn, sn1, 8);

                rad += 0.001;
            }
        }

    }
}

Runtime/iShape/FixFloat/FixSin.cs
namespace iShape.FixFloat {

    internal static class FixSin {

        internal static readonly long[] Map = {
            0,    // 0
            12,   // 1
            25,   // 2
            37,   // 3
            50,   // 4
            62,   // 5
            75,   // 6
            87,   // 7
            100,  // 8
            112,  // 9
            125,  // 10
            137,  // 11
            150,  // 12
            162,  // 13
            175,  // 14
            187,  // 15
            199,  // 16
            212,  // 17
            224,  // 18
            236,  // 19
            248,  // 20
            260,  // 21
            273,  // 22
            285,  // 23
            297,  // 24
            309,  // 25
            321,  // 26
            333,  // 27
            344,  // 28
            356,  // 29
            368,  // 30
            380,  // 31
            391,  // 32
            403,  // 33
            414,  // 34
            426,  // 35
            437,  // 36
            449,  // 37
            460,  // 38
            471,  // 39
            482,  // 40
            493,  // 41
            504,  // 42
            515,  // 43
            526,  // 44
            537,  // 45
            547,  // 46
            558,  // 47
            568,  // 48
            579,  // 49
            589,  // 50
            599,  // 51
            609,  // 52
            620,  // 53
            629,  // 54
            639,  // 55
            649,  // 56
            659,  // 57
            668,  // 58
            678,  // 59
            687,  // 60
            696,  // 61
            706,  // 62
            715,  // 63
            724,  // 64
            732,  // 65
            741,  // 66
            750,  // 67
            758,  // 68
            767,  // 69
            775,  // 70
            783,  // 71
            791,  // 72
            799,  // 73
            807,  // 74
            814,  // 75
            822,  // 76
            829,  // 77
            837,  // 78
            844,  // 79
            851,  // 80
            858,  // 81
            865,  // 82
            871,  // 83
            878,  // 84
            884,  // 85
            890,  // 86
            897,  // 87
            903,  // 88
            908,  // 89
            914,  // 90
            920,  // 91
            925,  // 92
            930,  // 93
            936,  // 94
            941,  // 95
            946,  // 96
            950,  // 97
            955,  // 98
            959,  // 99
            964,  // 100
            968,  // 101
            972,  // 102
            976,  // 103
            979,  // 104
            983,  // 105
            986,  // 106
            990,  // 107
            993,  // 108
            996,  // 109
            999,  // 110
            1001, // 111
            1004, // 112
            1006, // 113
            1008, // 114
            1010, // 115
            1012, // 116
            1014, // 117
            1016, // 118
            1017, // 119
            1019, // 120
            1020, // 121
            1021, // 122
            1022, // 123
            1022, // 124
            1023, // 125
            1023, // 126
            1023, // 127
            1024  // 128
        };
    }
}

Runtime/iShape/FixFloat/FixVec.cs
using System.Runtime.CompilerServices;
using Unity.Mathematics;

namespace iShape.FixFloat {

    /// <summary>
    /// Represents a fixed-point 2D vector using long integer coordinates.
    /// </summary>
    public readonly struct FixVec {

        /// <summary>
        /// The X coordinate of the vector.
        /// </summary>
        public readonly long x;
        
        /// <summary>
        /// The Y coordinate of the vector.
        /// </summary>
        public readonly long y;

        /// <summary>
        /// Constructs a new fixed-point vector with the specified X and Y coordinates.
        /// </summary>
        /// <param name="x">The X coordinate of the vector.</param>
        /// <param name="y">The Y coordinate of the vector.</param>
        public FixVec(long x, long y) {
            this.x = x;
            this.y = y;
        }

        /// <summary>
        /// Returns a zero vector (0, 0).
        /// </summary>
        public static FixVec Zero => new FixVec(0, 0);

        /// <summary>
        /// Packs the X and Y coordinates into a single long value.
        /// </summary>
        public long BitPack => (x << (FixNumber.MaxBits + 1)) + y;

        /// <summary>
        /// Converts the fixed-point vector to a float2 representation.
        /// </summary>
        /// <returns>A float2 representation of the vector.</returns>
        public float2 ToFloat2() {
            return new float2(x.ToFloat(), y.ToFloat());
        }

        /// <summary>
        /// Calculates the squared length of the vector.
        /// </summary>
        public long SqrLength => (x * x + y * y) >> FixNumber.FractionBits;

        /// <summary>
        /// Calculates the length of the vector.
        /// </summary>
        public long Length => (x * x + y * y).FastSquareRoot();

        /// <summary>
        /// Returns a normalized version of the vector.
        /// </summary>
        public FixVec Normalize {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get {
                long s = (1L << 30) / Length;
                long xx = (s * x) >> 20;
                long yy = (s * y) >> 20;
                
                return new FixVec(xx, yy);
            }
        }

        /// <summary>
        /// Returns a vector with half the length of the original vector.
        /// </summary>
        public FixVec Half => new FixVec(x >> 1, y >> 1);

        /// <summary>
        /// Returns an orthogonal vector, normalized and rotated 90 degrees clockwise or counterclockwise.
        /// </summary>
        /// <param name="clockwise">If true, returns a clockwise rotated vector. Otherwise, returns a counterclockwise rotated vector.</param>
        /// <returns>A normalized orthogonal vector.</returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public FixVec Ortho(bool clockwise) {
            return clockwise
                ? new FixVec(y, -x).Normalize
                : new FixVec(-y, x).Normalize;
        }
        
        /// <summary>
        /// Divides the vector coordinates by 2^count.
        /// </summary>
        /// <param name="count">The exponent to divide the vector coordinates by.</param>
        /// <returns>A new vector with divided coordinates.</returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public FixVec DivTwo(int count) {
            return new FixVec(x >> count, y >> count);
        }

        /// <summary>
        /// Adds two fixed-point vectors.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static FixVec operator +(FixVec left, FixVec right) {
            return new FixVec(left.x + right.x, left.y + right.y);
        }

        /// <summary>
        /// Subtracts two fixed-point vectors.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static FixVec operator -(FixVec left, FixVec right) {
            return new FixVec(left.x - right.x, left.y - right.y);
        }

        /// <summary>
        /// Multiplies a fixed-point vector by a scalar.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static FixVec operator *(FixVec left, long right) {
            long x = (left.x * right) >> FixNumber.FractionBits;
            long y = (left.y * right) >> FixNumber.FractionBits;
            return new FixVec(x, y);
        }
        
        /// <summary>
        /// Multiplies a scalar by a fixed-point vector.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static FixVec operator *(long left, FixVec right) {
            long x = (left * right.x) >> FixNumber.FractionBits;
            long y = (left * right.y) >> FixNumber.FractionBits;
            return new FixVec(x, y);
        }
        
        /// <summary>
        /// Calculates the dot product of two fixed-point vectors.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public long DotProduct(FixVec v)
        {
            long xx = (x * v.x) >> FixNumber.FractionBits;
            long yy = (y * v.y) >> FixNumber.FractionBits;
            return xx + yy;
        }

        /// <summary>
        /// Calculates the cross product of two fixed-point vectors.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public long CrossProduct(FixVec v)
        {
            long a = (x * v.y) >> FixNumber.FractionBits;
            long b = (y * v.x) >> FixNumber.FractionBits;

            return a - b;
        }
        
        /// <summary>
        /// Calculates the cross product of a fixed-point vector and a scalar.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public FixVec CrossProduct(long a)
        {
            long x0 = (a * y) >> FixNumber.FractionBits;
            long y0 = (a * x) >> FixNumber.FractionBits;

            return new FixVec(-x0, y0);
        }

        /// <summary>
        /// Calculates the squared distance between two fixed-point vectors.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public long SqrDistance(FixVec v)
        {
            return (this - v).SqrLength;
        }

        /// <summary>
        /// Calculates the middle point between two fixed-point vectors.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public FixVec Middle(FixVec v)
        {
            FixVec sum = this + v;
            return new FixVec(sum.x >> 1, sum.y >> 1);
        }
    }

}

Runtime/iShape/FixFloat/FixAngle.cs
using System.Runtime.CompilerServices;

namespace iShape.FixFloat {

    public static class FixAngle {

        private const long IndexMask = 256 - 1;
        private const long FullRoundMask = 1024 - 1;

        /// <summary>
        /// Returns the sine value for the specified angle.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long Sin(this long angle)
        {
            int quarter = (int)(angle & FullRoundMask) >> 8;
            int index = (int)(angle & IndexMask);

            switch (quarter)
            {
                case 0:
                    return Value(index);
                case 1:
                    return Value(256 - index);
                case 2:
                    return -Value(index);
                default:
                    return -Value(256 - index);
            }
        }
        
        /// <summary>
        /// Returns the cosine value for the specified angle.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long Cos(this long angle)
        {
            int quarter = (int)(angle & FullRoundMask) >> 8;
            int index = (int)(angle & IndexMask);

            switch (quarter)
            {
                case 0:
                    return Value(256 - index);
                case 1:
                    return -Value(index);
                case 2:
                    return -Value(256 - index);
                default:
                    return Value(index);
            }
        }
        
        /// <summary>
        /// Returns a rotation matrix for the specified angle.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static FixVec Rotator(this long angle)
        {
            int quarter = (int)(angle & FullRoundMask) >> 8;
            int index = (int)(angle & IndexMask);

            long sn;
            long cs;

            switch (quarter)
            {
                case 0:
                    sn = Value(index);
                    cs = Value(256 - index);
                    break;
                case 1:
                    sn = Value(256 - index);
                    cs = -Value(index);
                    break;
                case 2:
                    sn = -Value(index);
                    cs = -Value(256 - index);
                    break;
                default:
                    sn = -Value(256 - index);
                    cs = Value(index);
                    break;
            }

            return new FixVec(cs, sn);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static long Value(int index)
        {
            int i = index >> 1;
            if ((index & 1) == 1)
            {
                return (FixSin.Map[i] + FixSin.Map[i + 1]) >> 1;
            }
            else
            {
                return FixSin.Map[i];
            }
        }
        
        /// <summary>
        /// Converts a fixed-point radian value to a fixed-point angle value.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long RadToFixAngle(this long value) {
            return (value << 9) / FixNumber.PI;
        }
    }

}

Runtime/iShape/FixFloat/VecExtensions.cs
using System.Runtime.CompilerServices;
using Unity.Mathematics;

namespace iShape.FixFloat {

    public static class VecExtensions {

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static FixVec ToFixVec(this float2 vec)
        {
            return new FixVec(vec.x.ToFix(), vec.y.ToFix());
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float SqrLength(this float2 vec)
        {
            return math.lengthsq(vec);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float Length(this float2 vec)
        {
            return math.length(vec);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float2 Normalize(this float2 vec)
        {
            return math.normalize(vec);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float DotProduct(this float2 vec, float2 other)
        {
            return math.dot(vec, other);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float CrossProduct(this float2 vec, float2 other)
        {
            return vec.x * other.y - vec.y * other.x;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsClockwise(this float2 vec, float2 other)
        {
            return vec.x * other.y < vec.y * other.x;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float2 Ortho(this float2 vec, bool clockwise)
        {
            return clockwise ? new float2(vec.y, -vec.x).Normalize() : new float2(-vec.y, vec.x).Normalize();
        }
    }

}

Runtime/iShape/FixFloat/FixNumber.cs
using System.Runtime.CompilerServices;

namespace iShape.FixFloat {
    
    /// <summary>
    /// Provides methods and constants for working with fixed-point arithmetic.
    /// It supports numbers in the range 2^21 - 1 to -2^21 + 1
    /// with a precision of 1/1024, with the ideal range being 10,000,000 to
    /// -10,000,000 with a precision of 0.01.
    /// </summary>
    public static class FixNumber
    {
        
        /// <summary>
        /// The number of bits representing the fractional part of the fixed-point number.
        /// 1 / 1024 ~ 0.001 = 1
        /// 256 = 0.25
        /// 1024 = 1
        /// (1024 + 512) = 1.5
        /// (2048 + 256) = 2.25  
        /// </summary>
        public const int FractionBits = 10;
        
        /// <summary>
        /// The maximum number of bits for the integer part of the fixed-point number. 31
        /// </summary>
        public const int MaxBits = (64 >> 1) - 1;
        
        /// <summary>
        /// The maximum value of the fixed-point number. 2^31 - 1
        /// </summary>
        public const long Max = (1L << MaxBits) - 1;
        
        /// <summary>
        /// The minimum value of the fixed-point number. -2^31 + 1
        /// </summary>
        public const long Min = -Max;

        /// <summary>
        /// The fixed-point representation of 1.0.
        /// </summary>
        public const long Unit = 1L << FractionBits;
        
        /// <summary>
        /// The fixed-point representation of 0.5.
        /// </summary>
        public const long Half = 1L << (FractionBits - 1);

        /// <summary>
        /// The fixed-point representation of PI.
        /// </summary>
        public const long PI = 3217;

        /// <summary>
        /// Divides a fixed-point number by another number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long Div(this long number, long value)
        {
            return (number << FractionBits) / value;
        }

        /// <summary>
        /// Multiplies a fixed-point number by another number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long Mul(this long number, long value)
        {
            return (number * value) >> FractionBits;
        }

        /// <summary>
        /// Calculates the square of a fixed-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long Sqr(this long number)
        {
            return (number * number) >> FractionBits;
        }

        /// <summary>
        /// Calculates the square root of a fixed-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long Sqrt(this long number)
        {
            return (number << FractionBits).FastSquareRoot();
        }

        /// <summary>
        /// Converts a fixed-point number to a double-precision floating-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static double ToDouble(this long number)
        {
            return (double)number / Unit;
        }

        /// <summary>
        /// Converts a fixed-point number to a single-precision floating-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float ToFloat(this long number)
        {
            return (float)number / Unit;
        }

        /// <summary>
        /// Converts a fixed-point number to an integer.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int ToInt(this long number)
        {
            return (int)(number >> FractionBits);
        }

        /// <summary>
        /// Converts an integer to a fixed-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long FromInt(long value)
        {
            return value << FractionBits;
        }

        /// <summary>
        /// Converts a double-precision floating-point number to a fixed-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long ToFix(this double value)
        {
            return (long)(value * Unit);
        }

        /// <summary>
        /// Converts a floating-point number to a fixed-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long ToFix(this float value)
        {
            return (long)(value * Unit);
        }

        /// <summary>
        /// Converts a integer number to a fixed-point number.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static long ToFix(this int value)
        {
            return (long)value << FractionBits;
        }
    }

    internal static class Int64FastRoot
    {

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal static long FastSquareRoot(this long number) {
            long a = (long)System.Math.Sqrt(number);
            long b = a + 1;

            return b * b > number ? a : b;
        }
    }
}